### ISSUE TITLE
Public Search (Case): Allow Specific Date Range

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8454]
As a member of the public, so that I can narrow my Case search, I need to be able to specify a date range for my search criteria.

Currently, the public-facing Case search allows users to only specify the year in which the case was filed. To allow for narrower searches and to bring it in line with the models used for Order and Opinion searches, we would like to replace this with a custom date range function specifying valid start and end dates for the filing date of the Cases in question.

## Acceptance Criteria
* On the Case search UI, the Year Filed (from and to) criteria is replaced with Date Filed (from and to) that requires a specific date range
* Date Filed format/validation is consistent with other date range fields across DAWSON
* When populated, only cases with a Filed Date within the specified date range are retrieved as search results
* Date range works in conjunction with other search criteria

## Notes
* Consistent = ?

## Story Definition of Ready (updated on 12/23/22)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Is framed in business/user need, the value has been addressed.
- [ ] Includes acceptance criteria
- [ ] Has been refined
- [ ] Pre conditions have been satisfied. 
**Process:** 
Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written, including visual tests for newly added PDFs
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by swongCO** [comment]
Notes from pre-refinement:
Date component has been created; please reuse. Not the one in Order Search.

**Comment by katiecissell** [comment]
Add the new date picker component below the country/state fields.
![image.png](https://images.zenhubusercontent.com/63ecf8daa45ad187539a9d61/bb195a41-2b8c-41f1-8de4-5e1ec95935c5)

- The hint text should be moved out of the placeholder area and put below the label. (There will be a design debt ticket to follow this to update the other date picker components to match)

Validation:
Start date component:
- Should disable dates after today
- Should disable dates after selected end date
- Should only allow numbers and /’s

End date component:
- Should disable dates after today
- Should disable dates before selected start date
- Should only allow numbers and /’s

Validation
- Only Petitioner name is a required field
- User can enter just a start date, it will search from the start date to today. 
- User can enter just a end date, it will search  all dates before the end date
- If date is entered in a format other than mm/dd/yyyy then display a message when search is clicked. 
- - Display an alert "Please correct the following errors:" "Format date as MM/DD/YYYY"
- and inline validation message "Format date as MM/DD/YYYY"

End date
- If date is typed in that is beyond today’s date
- - display an alert "Please correct the following errors:" "End date cannot be in the future."
- and inline validation message "End date cannot be in the future. "
- If end date is typed in that is prior to start date
- - display an alert  "Please correct the following errors:" "End date cannot be prior to start date."
- and inline validation message "End date cannot be prior to start date. "

 Start Date
- If date is typed in that is beyond today’s date
- display an alert "Please correct the following errors:" "End date cannot be in the future."
- and inline validation message "End date cannot be in the future. "
- If start date is typed in that is after end date
- display an alert  "Please correct the following errors:" "Start date cannot be after end date."
- and inline validation message "Start date cannot be after end date. "
![image.png](https://images.zenhubusercontent.com/63ecf8daa45ad187539a9d61/8f98ba79-536b-4c26-8395-fdc77c96cb83)


Mobile: Date fields should stack and span the width to match the other field widths.
![image.png](https://images.zenhubusercontent.com/63ecf8daa45ad187539a9d61/4bbaf793-55d8-48e3-878a-095aeea0faa0)

[Figma File](https://www.figma.com/file/XxBdxjXtbbeXsUuU4N3yud/USTC-%7C-Public-Search-(Case)%3A-Allow-Specific-Date-Range-%7C-10052?type=design&node-id=32%3A3251&t=xjB2WCYp3OrvD2j8-1)

**Comment by ttlenard** [comment]
## Test Cases

#### 1) Public non-logged in user navigates to the Case Search page; there is a new Date filed start date and a Date filed end date field.

- Navigate to the DAWSON Case Search page
- Review the Search by Name section of the page

Expected Results:
- [x] **NEW FUNCTIONALITY** - There are two new fields present; Date filed start date and Date filed end date
- [x] **NEW FUNCTIONALITY** - Both fields have a calendar icon next to the fields
- [x] **NEW FUNCTIONALITY** - After clicking on the calendar icon for the start date field, dates in the future are not selectable
- [x] **NEW FUNCTIONALITY** - After clicking on the calendar icon for the end date field, dates in the future are not selectable
- [x] **NEW FUNCTIONALITY** - There is no helper text within the date fields


#### 2) Public non-logged in user inputs a Petitioner name, adds 2 valid dates in the new fields by selecting the date from the calendar icon, clicks search; Results list displays results appropriately by only displaying cases in the date range that was added.

- Continuing from the previous test
- Input a Petitioner name
- Add in valid dates (start date not in the future, end date after the start date, in the MM/DD/YYYY format)
- Click Search

Expected Results:
- [x] The results list includes appropriate cases that contain the name that you input in the name field
- [x] **NEW FUNCTIONALITY** - The results list includes appropriate cases that were filed between the dates that you input in the date fields.


#### 3) Public non-logged in user inputs a Petitioner name, State, valid dates, and clicks search; Results list displays results appropriately by only displaying cases in the appropriate state and date range that was added.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in a state
- Add in valid dates (start date not in the future, end date after the start date, in the MM/DD/YYYY format)
- Click Search

Expected Results:
- [x] The results list includes appropriate cases that contain the name that you input in the name field
- [x] The results list includes appropriate cases that contain the state that you put in the state field
- [x] **NEW FUNCTIONALITY** - The results list includes appropriate cases that were filed between the dates that you input in the date fields.


#### 4) Public non-logged in user inputs a Petitioner name, adds invalid date format; Clicks Search; receives validation error.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in an valid date format (ex: 5/9/21)
- Click Search

Expected Results:
- [x] **NEW FUNCTIONALITY** - A validation error is displayed at the top of the page that states, "Please correct the following errors:" "Format date as MM/DD/YYYY"
- [x] **NEW FUNCTIONALITY** - An inline validation error is displayed that states "Format date as MM/DD/YYYY"

*Repeat this test by adding in other variations of invalid date formats in BOTH the start and end date fields.


#### 5) Public non-logged in user inputs a Petitioner name, adds a start date in the future; Clicks Search; receives validation error.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in a start date that is in the future
- Click Search

Expected Results:
- [x] **NEW FUNCTIONALITY** - A validation error is displayed at the top of the page that states, "Start date cannot be in the future"
- [x] **NEW FUNCTIONALITY** - An inline validation error is displayed that states "Start date cannot be in the future"


#### 6) Public non-logged in user inputs a Petitioner name, adds a start date that is after the end date; Clicks Search; receives validation error.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in a start date that is after the end date
- Click Search

Expected Results:
- [x] **NEW FUNCTIONALITY** - A validation error is displayed at the top of the page that states, "Start date cannot be after end date"
- [x] **NEW FUNCTIONALITY** - An inline validation error is displayed that states "Start date cannot be after end date"


#### 7) Public non-logged in user inputs a Petitioner name, adds an end date in the future; Clicks Search; receives validation error.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in a end date that is in the future
- Click Search

Expected Results:
- [x] **NEW FUNCTIONALITY** - A validation error is displayed at the top of the page that states, "End date cannot be in the future"
- [x] **NEW FUNCTIONALITY** - An inline validation error is displayed that states "End date cannot be in the future"



#### 8) Public non-logged in user inputs a Petitioner name, adds an end date that is before the start date; Clicks Search; receives validation error.

- Continuing from the previous test
- Clear the search
- Input a petitioner name
- Add in a end date that is prior to the start date
- Click Search

Expected Results:
- [x] **NEW FUNCTIONALITY** - A validation error is displayed at the top of the page that states, "End date cannot be prior to start date"
- [x] **NEW FUNCTIONALITY** - An inline validation error is displayed that states "End date cannot be prior to start date"


#### 9) New date picker fields display appropriately when viewing on a mobile device.

- Navigate to the DAWSON case search page on a mobile device
- Review the Search by Name section of the page

Expected results:
- [x] The new date filed start date and date filed end date fields are displayed in a manner that is easy to view on a mobile device
- [x] All other functionality as indicated in test cases 1-8 function on mobile as well.


#### 10) Repeat tests 1-9 as other users.

- [x] Tests 1-9 pass when searching cases as a Private Practitioner
- [x] Tests 1-9 pass when searching cases as an IRS Practitioner
- [x] Tests 1-9 pass when searching cases as the Public Terminal user
- [x] Tests 1-9 pass when searching cases as an Internal Court user (assuming that this story get's extended to Internal Court staff as well as public user???)

**Comment by rgoshen** [comment]
@cholly75 and @ttlenard good news, from what I can see when I implement this, it should also implement for internal users as well.  Regardless of internal or external user, it is the same component.

**Comment by ttlenard** [comment]
Testing Feedback. @codyseibert 

- I notice that the results aren't listed newest to oldest by docket number.  (It mostly is, but not quite. - - There seems to be like 1 docket number off when I do a search.)
- I found the original story with the A/C that says newest to oldest by docket number: https://github.com/flexion/ef-cms/issues/2828
- Screen grabs show the out of order docket numbers for two searches I did.  If you expand the dates to get more results, it's the same and there are more cases out of order.

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/f676425f-1b6c-40d6-bf9b-b70fc4c3ad0c)

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/9de58fba-8f06-412b-bfd4-84ccd1b250fc)
